### PR TITLE
fix Issue 20270 - [REG2.087] Deadlock in garbage collection when runn…

### PR DESCRIPTION
**Pull request by rainers**
…ing processes in parallel

start scan threads while the world isn't suspended

### CODE REVIEW

**Comment by thewilsonator** [comment]
Is it possible to test?

**Comment by CyberShadow** [comment]
Thank you! Confirmed the fix on my machine.

> Is it possible to test?

There is a small test program in the bug report. It should be suitable to be added to the test suite.

**Comment by rainers** [comment]
> There is a small test program in the bug report. It should be suitable to be added to the test suite.

It depends on phobos which is not allowed in druntime. The root cause is much more basic, but I'm not sure it is easy to reproduce.

**Comment by CyberShadow** [comment]
I'll try to create a smaller test case.

**Comment by CyberShadow** [comment]
Reduced test case:

```d
import core.stdc.stdlib : exit;
import core.sys.posix.unistd : fork;
import core.thread : Thread;

void main()
{
	foreach (t; 0 .. 10)
		new Thread({
			foreach (n; 0 .. 100)
			{
				foreach (x; 0 .. 100)
					new ubyte[x];
				auto f = fork();
				assert(f >= 0);
				if (f == 0)
					exit(0);
			}
		}).start();
}
```

**Comment by CyberShadow** [comment]
I think I found a new bug while trying to create a test case:

```d
import core.stdc.stdlib : exit;
import core.sys.posix.sys.wait;
import core.sys.posix.unistd : fork;
import core.thread : Thread;

void main()
{
	foreach (t; 0 .. 10)
		new Thread({
			foreach (n; 0 .. 100)
			{
				foreach (x; 0 .. 100)
					new ubyte[x];
				auto f = fork();
				assert(f >= 0);
				if (f == 0)
					exit(0);
				else
					waitpid(f, null, 0);
			}
		}).start();
}
```

This program has a high chance of never exiting, because it is waiting on forked processes which are deadlocked. The forked processes have a single thread in `core.internal.spinlock.SpinLock.yield` (called by `Gcx.removeRange`).

It looks like the GC lock was not cleared after a fork?

**Comment by CyberShadow** [comment]
> Reduced test case:

For this bug / PR:

```d
import core.sys.posix.sys.wait : waitpid;
import core.sys.posix.unistd : fork, _exit;
import core.thread : Thread;

void main()
{
	foreach (t; 0 .. 10)
		new Thread({
			foreach (n; 0 .. 100)
			{
				foreach (x; 0 .. 100)
					new ubyte[x];
				auto f = fork();
				assert(f >= 0);
				if (f == 0)
					_exit(0);
				else
					waitpid(f, null, 0);
			}
		}).start();
}
```

**Comment by CyberShadow** [comment]
> I think I found a new bug while trying to create a test case:

I meant to call `_exit` instead of `exit`. The latter calls finalizers incl. GC destruction, but the fork might have happened while the GC lock was being held. This is a very separate issue. Probably, it would be good to keep the GC working after a fork, so it would need to clear its lock in its `atfork` handler.

**Comment by rainers** [comment]
Thanks, I added the test case. 

> The latter calls finalizers incl. GC destruction, but the fork might have happened while the GC lock was being held. This is a very separate issue. Probably, it would be good to keep the GC working after a fork, so it would need to clear its lock in its `atfork` handler.

I think there are more issues in druntime like that. For example the chain of Threads all point to dead threads in the fork-child, so joining these will fail, too.

**Comment by CyberShadow** [comment]
> I think there are more issues in druntime like that. For example the chain of Threads all point to dead threads in the fork-child, so joining these will fail, too.

Hmm, probably we should try to fix them one by one.

What do you think about https://github.com/dlang/druntime/pull/2817 ?

**Comment by rainers** [comment]
@CyberShadow This is working now, unfortunately a day too late for the release.

**Comment by wilzbach** [comment]
@rainers FYI: the master-stable branchoff hasn't happened yet.

**Comment by rainers** [comment]
@wilzbach I meant the 2.088.1 release, this PR is targeting stable.

File: src/gc/impl/conservative/gc.d [hunk 1]
```diff
@@ -2871,8 +2871,12 @@ struct Gcx
             if (scanThreadData[idx].tid != scanThreadData[idx].tid.init)
                 startedThreads++;
 
+        version (Windows)
+            alias allThreadsDead = thread_DLLProcessDetaching;
+        else
+            enum allThreadsDead = false;
```
Comment (CyberShadow) [inline src/gc/impl/conservative/gc.d]: So what happened here? A DLL with running threads was unloaded, terminating them (causing them to segfault)?

Comment (rainers) [reply src/gc/impl/conservative/gc.d]: No, the process terminates while threads are still running. The threads are just terminated as well without `DllMain(DLL_THREAD_DETACH)` being called. `DllMain(DLL_PROCESS_DETACH) is still ìnvoked, though.

**Comment by rainers** [comment]
Rebased. Would be good to have this in the coming release as the Visual D installer is hit by the non-termination issue, too.

**Pull request merged**